File: 2024/d1.php
<?php declare(strict_types=1);

$lists = [
    'left' => [],
    'right' => [],
];

$file = file(filename: __DIR__ . '/d1.txt');

foreach ($file as $key => $line) {
    $ids = explode(separator: '   ', string: $line);

    $lists['left'][] = $ids[0];
    $lists['right'][] = $ids[1];
}

foreach ($lists as $key => $value) {
    sort(array: $lists[$key]);
}

assert(assertion: count(value: $lists['left']) === count(value: $lists['right']));

$totalDistance = 0;

foreach ($lists['left'] as $key => $value) {
    $sum = $value - $lists['right'][$key];

    if ($sum < 0) $sum = -$sum;

    $totalDistance += $sum;
}

echo "Total distance: $totalDistance\n";

/**
 * Part two
 */

$similarityScore = 0;

foreach ($lists['left'] as $value) {
    $times = count(value: array_filter(
        array: $lists['right'],
        callback: static fn (int $id): bool => $id == $value
    ));

    $similarityScore += $value * $times;
}

echo "Similarity score: $similarityScore\n";

File: 2024/elixir/d1.exs
lists = %{
  "left" => [],
  "right" => []
}

file_lines = File.stream!("/data/d1.txt")
  |> Enum.to_list()

lists =
  for line <- file_lines, reduce: lists do
    acc ->
      ids = String.split(line, "   ")
      acc
      |> Map.update!("left", &([hd(ids) | &1]))
      |> Map.update!("right", &([Enum.at(ids, 1) | &1]))
  end

lists = Map.update!(lists, "left", &Enum.reverse/1)
lists = Map.update!(lists, "right", &Enum.reverse/1)


IO.inspect(lists)
